### PR TITLE
Issue #5088 Review ContextHandler locking

### LINKED ISSUE #5088: Review ContextHandler locking
Spawned by #5083.

**Jetty version**
9.4.x

**Description**
`ContextHandler` uses a mix of `synchronized` modifiers and `synchronized (this)` inside methods that is difficult to reconcile.

Furthermore, it has a `setClassLoader()` method that allow to arbitrarily change the `_classLoader` field but the rest of the code does not protect with a lock this field, nor it reads the field only once.
For example, `loadClass()`, `handle()`, `addListener()`, etc. reference the `_classLoader` field multiple times and the field may change between the references.

Also, `handle()` could have an improved logic.
In `handle()`, in the `finally` block, rather than checking on `oldClassLoader` the check should be done symmetrically on `_classLoader` like the code before the `try` does and like `doStart()` does.

Also, `setAvailable()` is locked, but `isAvailable()` is not. In general, `_availability` is not correctly locked.

### PR DESCRIPTION
**Pull request by gregw**
The locking was primarily as a memory guard for the availability status, which was already volatile.
Have instead using an AtomicReference with a simple state machine layered on top of start/stop lifecycle.
There was also protection for AttributesMap, which is no longer needed as AttributesMap is now concurrent.

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/Request.java [hunk 1]
```diff
@@ -1805,7 +1805,7 @@ public void setMetaData(org.eclipse.jetty.http.MetaData.Request request)
         }
         else if (encoded.startsWith("/"))
         {
-            path = (encoded.length() == 1) ? "/" : URIUtil.canonicalPath(URIUtil.decodePath(encoded));
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:1808]: strictly speaking this has nothing to do with contexthandler locking.

File: jetty-server/src/main/java/org/eclipse/jetty/server/Request.java [hunk 2]
```diff
@@ -1805,7 +1805,7 @@ public void setMetaData(org.eclipse.jetty.http.MetaData.Request request)
...
+            path = (encoded.length() == 1) ? "/" : URIUtil.canonicalPath(uri.getDecodedPath());
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:1808]: oops this is an unrelated change that accidentally slipped in.  
The `uri.getDecodedPath()` method  does exactly `URIUtil.decodePath(encoded)`, but caches the results so they can be reused if needed again.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java [hunk 3]
```diff
@@ -777,7 +778,23 @@ public boolean isShutdown()
...
+        {
+            Availability availability = _availability.get();
+            switch (availability)
+            {
+                case STOPPED:
+                    return new FutureCallback(new IllegalStateException(getState()));
+                case STARTING:
+                case AVAILABLE:
+                    if (!_availability.compareAndSet(availability, Availability.SHUTDOWN))
+                        continue;
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:796]: If UNAVAILABLE, should the state go to SHUTDOWN?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java [hunk 4]
```diff
@@ -796,12 +813,29 @@ public boolean isAvailable()
...
+            {
+                Availability availability = _availability.get();
+                switch (availability)
+                {
+                    case STARTING:
+                    case AVAILABLE:
+                        if (!_availability.compareAndSet(availability, Availability.UNAVAILABLE))
+                            continue;
+                        break;
+                    default:
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:856]: Maybe should throw an ISE if you try to set availability whilst shutdown or stopped?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java [hunk 5]
```diff
@@ -230,10 +231,10 @@ public static void setServerInfo(String serverInfo)
 
     public enum Availability
     {
-        UNAVAILABLE, STARTING, AVAILABLE, SHUTDOWN,
+        STOPPED, STARTING, AVAILABLE, UNAVAILABLE, SHUTDOWN,
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java]: Can we have a comment about the meanings of the different states, particularly UNAVAILABLE vs STOPPED, which is a new state.

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java [hunk 6]
```diff
@@ -1434,7 +1434,7 @@ protected void stopContext() throws Exception
                 setClassLoader(null);
             }
 
-            setAvailable(true);
```
Comment (janbartel) [inline jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java:1437]: Need an explanation of this change as on the face of it, it is the exact opposite of the original intent.

**Review by janbartel — COMMENTED** [review]
I wonder if this should be done in 9 or 10? We are changing the signature of exposed methods, albeit only  via removal of the "synchronized" keyword, but tooling may flag this as an api compatibility problem for a point release?

File: jetty-server/src/main/java/org/eclipse/jetty/server/Request.java (hunk 1 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:1808]: sorry, but it is little :)

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java (hunk 4 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:856]: I think perhaps only if trying to set available==true, as all other states are equivalent to available==false

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java (hunk 6 above)
Comment (gregw) [reply jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java:1437]: I can't really explain.  It looks to me that it was just plain wrong, but didn't matter because we were stopping already.
I think it actually dates from when availability was just in the webapp and it had to clear unavailable status   This is no longer required, so I will try without anything here.

**Review by janbartel — APPROVED** [review]

File: jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilCanonicalPathTest.java [hunk 7]
```diff
@@ -114,6 +114,13 @@
                 // paths with encoded segments should remain encoded
                 // canonicalPath() is not responsible for decoding characters
                 {"%2e%2e/", "%2e%2e/"},
+                {"/%2e%2e/", "/%2e%2e/"},
+
+                // paths with parameters are not elided
+                // canonicalPath() is not responsible for decoding characters
+                {"/foo/.;/bar", "/foo/.;/bar"},
+                {"/foo/..;/bar", "/foo/..;/bar"},
+                {"/foo/..;/..;/bar", "/foo/..;/..;/bar"},
```
Comment (joakime) [inline jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilCanonicalPathTest.java:123]: This is an unrelated change.

Comment (gregw) [reply jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilCanonicalPathTest.java:123]: Yes, see note explaining it above.

Comment (gregw) [reply jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilCanonicalPathTest.java:123]: Oh no it's another unrelated change!   Double ooops

Comment (joakime) [reply jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilCanonicalPathTest.java:123]: These are useful and good changes.
Just would have liked to see them in a different commit explaining why they were made.
Right now they seem to be related to the change for commit 36b2a4bbc9e628185747d32c4044067bb2b15419 (Issue #5088 Review ContextHandler Locking), which will be super confusing in the future.

Comment (gregw) [reply jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilCanonicalPathTest.java:123]: yep - my bad!
